msal_extensions/token_cache.py

```diff
@@ -1,156 +1,119 @@
 """Generic functions and types for working with a TokenCache that is not platform specific."""
 import os
-import sys
 import warnings
 import time
 import errno
-import msal
-from .cache_lock import CrossPlatLock
+import logging
 
-if sys.platform.startswith('win'):
-    from .windows import WindowsDataProtectionAgent
-elif sys.platform.startswith('darwin'):
-    from .osx import Keychain
+import msal
 
-def _mkdir_p(path):
-    """Creates a directory, and any necessary parents.
+from .cache_lock import CrossPlatLock
+from .persistence import (
+    _mkdir_p, FilePersistence,
+    DataProtectionPersistence, KeychainPersistence, LibsecretPersistence)
 
-    This implementation based on a Stack Overflow question that can be found here:
-    https://stackoverflow.com/questions/600268/mkdir-p-functionality-in-python
 
-    If the path provided is an existing file, this function raises an exception.
-    :param path: The directory name that should be created.
-    """
-    try:
-        os.makedirs(path)
-    except OSError as exp:
-        if exp.errno == errno.EEXIST and os.path.isdir(path):
-            pass
-        else:
-            raise
+logger = logging.getLogger(__name__)
 
+class PersistedTokenCache(msal.SerializableTokenCache):
+    """A token cache using given persistence layer, coordinated by a file lock."""
 
-class FileTokenCache(msal.SerializableTokenCache):
-    """Implements basic unprotected SerializableTokenCache to a plain-text file."""
-    def __init__(self,
-                 cache_location,
-                 lock_location=None):
-        super(FileTokenCache, self).__init__()
-        self._cache_location = cache_location
-        self._lock_location = lock_location or self._cache_location + '.lockfile'
+    def __init__(self, persistence, lock_location=None):
+        super(PersistedTokenCache, self).__init__()
+        self._lock_location = (
+            os.path.expanduser(lock_location) if lock_location
```

Comment:
Just to confirm - the location of the lockfile is the same as the location of the cache file (or dummy cache file)..

Reply:
By "same location", did you mean "same folder/directory"? Indeed, by default we just add a ".lockfile" suffix to the cache filename, to form the lock filename (see line 25 below). Historically the API surface also accepts a customized lockfile location, and perhaps that is a feature also available in .Net/Java too, although I doubt how many people actually use that behavior.